Code quality check fails because of new flake8 F824 error

The code quality check is failing with:
```console
$ flake8 sympy
sympy/integrals/laplace.py:80:5: F824 `global _LT_level` is unused: name is never assigned in scope
sympy/integrals/meijerint.py:591:5: F824 `global _dummies` is unused: name is never assigned in scope
sympy/printing/llvmjitcode.py:183:9: F824 `global link_names` is unused: name is never assigned in scope
sympy/printing/llvmjitcode.py:266:9: F824 `global exe_engines` is unused: name is never assigned in scope
sympy/simplify/fu.py:656:5: F824 `global _ROOT2` is unused: name is never assigned in scope
sympy/simplify/fu.py:656:5: F824 `global _ROOT3` is unused: name is never assigned in scope
sympy/simplify/fu.py:656:5: F824 `global _invROOT3` is unused: name is never assigned in scope
sympy/simplify/fu.py:1778:5: F824 `global _ROOT2` is unused: name is never assigned in scope
sympy/simplify/fu.py:1778:5: F824 `global _ROOT3` is unused: name is never assigned in scope
sympy/simplify/fu.py:1778:5: F824 `global _invROOT3` is unused: name is never assigned in scope
sympy/testing/tests/diagnose_imports.py:[12](https://github.com/sympy/sympy/actions/runs/14154391175/job/39651743753?pr=27827#step:9:13)6:9: F824 `global options` is unused: name is never assigned in scope
sympy/testing/tests/diagnose_imports.py:126:9: F824 `global sorted_messages` is unused: name is never assigned in scope
sympy/testing/tests/diagnose_imports.py:[14](https://github.com/sympy/sympy/actions/runs/14154391175/job/39651743753?pr=27827#step:9:15)6:9: F824 `global options` is unused: name is never assigned in scope
sympy/testing/tests/diagnose_imports.py:146:9: F824 `global symbol_definers` is unused: name is never assigned in scope
sympy/utilities/timeutils.py:55:9: F824 `global _do_timings` is unused: name is never assigned in scope
```
I think this is because flake 7.2.0 was released yesterday:

https://pypi.org/project/flake8/#history

To see these errors you need to update to the latest flake8 version e.g.:
```
pip install -U flake8
```

The error F824 is not yet listed in the flake8 docs:

https://flake8.pycqa.org/en/latest/user/error-codes.html

I think that the error is valid and basically just means that the global statement is being used in situations where it is not needed. We could disable this error code but I would rather enable it in the ruff config instead and have these all be fixed. We should add this with ruff because in general we want to move from using flake8 to using ruff so any time flake8 picks up on something valid we should add that rule to the ruff config.

The corresponding ruff rule seems to be PLW0602:

https://docs.astral.sh/ruff/rules/global-variable-not-assigned/

With this diff ruff will check this:
```diff
diff --git a/pyproject.toml b/pyproject.toml
index f6075eba23..610004d9f1 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -52,6 +52,7 @@ lint.select = [
     "SIM101",
     "SLOT",
     "TRY002",
+    "PLW0602",
 ]
 
 # Ignore rules that currently fail on the SymPy codebase
```
Then we get
```console
$ $ ruff check sympy
sympy/integrals/laplace.py:80:12: PLW0602 Using global for `_LT_level` but no assignment is done
   |
78 | def _debug(text):
79 |     from sympy import SYMPY_DEBUG
80 |     global _LT_level
   |            ^^^^^^^^^ PLW0602
81 |
82 |     if SYMPY_DEBUG:
   |

sympy/integrals/meijerint.py:591:12: PLW0602 Using global for `_dummies` but no assignment is done
    |
589 |     it globally.
590 |     """
591 |     global _dummies
    |            ^^^^^^^^ PLW0602
592 |     if not (name, token) in _dummies:
593 |         _dummies[(name, token)] = Dummy(name, **kwargs)
    |

sympy/printing/llvmjitcode.py:183:16: PLW0602 Using global for `link_names` but no assignment is done
    |
181 |     def _create_function_base(self):
182 |         """Create function with name and type signature"""
183 |         global link_names, current_link_suffix
    |                ^^^^^^^^^^ PLW0602
184 |         default_link_name = 'jit_func'
185 |         current_link_suffix += 1
    |

sympy/printing/llvmjitcode.py:266:16: PLW0602 Using global for `exe_engines` but no assignment is done
    |
265 |     def _compile_function(self, strmod):
266 |         global exe_engines
    |                ^^^^^^^^^^^ PLW0602
267 |         llmod = llvm.parse_assembly(strmod)
    |

sympy/simplify/fu.py:656:12: PLW0602 Using global for `_ROOT2` but no assignment is done
    |
655 |     """
656 |     global _ROOT2, _ROOT3, _invROOT3
    |            ^^^^^^ PLW0602
657 |     if _ROOT2 is None:
658 |         _roots()
    |

sympy/simplify/fu.py:656:20: PLW0602 Using global for `_ROOT3` but no assignment is done
    |
655 |     """
656 |     global _ROOT2, _ROOT3, _invROOT3
    |                    ^^^^^^ PLW0602
657 |     if _ROOT2 is None:
658 |         _roots()
    |

sympy/simplify/fu.py:656:28: PLW0602 Using global for `_invROOT3` but no assignment is done
    |
655 |     """
656 |     global _ROOT2, _ROOT3, _invROOT3
    |                            ^^^^^^^^^ PLW0602
657 |     if _ROOT2 is None:
658 |         _roots()
    |

sympy/simplify/fu.py:1778:12: PLW0602 Using global for `_ROOT2` but no assignment is done
     |
1776 |     >>> trig_split(-sqrt(6)*cos(x), sqrt(2)*sin(x)*sin(y), two=True)
1777 |     """
1778 |     global _ROOT2, _ROOT3, _invROOT3
     |            ^^^^^^ PLW0602
1779 |     if _ROOT2 is None:
1780 |         _roots()
     |

sympy/simplify/fu.py:1778:20: PLW0602 Using global for `_ROOT3` but no assignment is done
     |
1776 |     >>> trig_split(-sqrt(6)*cos(x), sqrt(2)*sin(x)*sin(y), two=True)
1777 |     """
1778 |     global _ROOT2, _ROOT3, _invROOT3
     |                    ^^^^^^ PLW0602
1779 |     if _ROOT2 is None:
1780 |         _roots()
     |

sympy/simplify/fu.py:1778:28: PLW0602 Using global for `_invROOT3` but no assignment is done
     |
1776 |     >>> trig_split(-sqrt(6)*cos(x), sqrt(2)*sin(x)*sin(y), two=True)
1777 |     """
1778 |     global _ROOT2, _ROOT3, _invROOT3
     |                            ^^^^^^^^^ PLW0602
1779 |     if _ROOT2 is None:
1780 |         _roots()
     |

sympy/testing/tests/diagnose_imports.py:126:16: PLW0602 Using global for `options` but no assignment is done
    |
125 |     def msg(msg, *args):
126 |         global options, sorted_messages
    |                ^^^^^^^ PLW0602
127 |         if options.by_process:
128 |             print(msg % args)
    |

sympy/testing/tests/diagnose_imports.py:126:25: PLW0602 Using global for `sorted_messages` but no assignment is done
    |
125 |     def msg(msg, *args):
126 |         global options, sorted_messages
    |                         ^^^^^^^^^^^^^^^ PLW0602
127 |         if options.by_process:
128 |             print(msg % args)
    |

sympy/testing/tests/diagnose_imports.py:146:16: PLW0602 Using global for `options` but no assignment is done
    |
145 |         Keeps the semantics of __import__ unchanged."""
146 |         global options, symbol_definers
    |                ^^^^^^^ PLW0602
147 |         caller_frame = inspect.getframeinfo(sys._getframe(1))
148 |         importer_filename = caller_frame.filename
    |

sympy/testing/tests/diagnose_imports.py:146:25: PLW0602 Using global for `symbol_definers` but no assignment is done
    |
145 |         Keeps the semantics of __import__ unchanged."""
146 |         global options, symbol_definers
    |                         ^^^^^^^^^^^^^^^ PLW0602
147 |         caller_frame = inspect.getframeinfo(sys._getframe(1))
148 |         importer_filename = caller_frame.filename
    |

sympy/utilities/timeutils.py:55:16: PLW0602 Using global for `_do_timings` but no assignment is done
   |
53 | def timethis(name):
54 |     def decorator(func):
55 |         global _do_timings
   |                ^^^^^^^^^^^ PLW0602
56 |         if name not in _do_timings:
57 |             return func
   |

Found 15 errors.
```
That seems to be the same errors that flake8 picks up on.

In many cases it might be better not to use `global` variables at all or to use `@cache` or something. Generally it is a bad idea to use `global` and historically it has been used badly in the SymPy codebase. It is typically not thread-safe and there other problems with it.

Rather than adding the PLW0602 rule specifically we could consider adding ruff's whole PLW block of rules:

https://docs.astral.sh/ruff/rules/#warning-plw

I tried that:
```diff
diff --git a/pyproject.toml b/pyproject.toml
index f6075eba23..97ccb42afd 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -52,6 +52,7 @@ lint.select = [
     "SIM101",
     "SLOT",
     "TRY002",
+    "PLW",
 ]
 
 # Ignore rules that currently fail on the SymPy codebase
```
That gives the following:
```console
ruff check sympy --statistics
510	PLW2901	[ ] redefined-loop-name
 42	PLW0603	[ ] global-statement
 20	PLW0642	[ ] self-or-cls-assignment
 15	PLW0602	[ ] global-variable-not-assigned
  7	PLW0211	[ ] bad-staticmethod-argument
  7	PLW3301	[ ] nested-min-max
  3	PLW0120	[*] useless-else-on-loop
  3	PLW0128	[ ] redeclared-assigned-name
  2	PLW0127	[ ] self-assigning-variable
  1	PLW0133	[ ] useless-exception-statement
Found 610 errors.
[*] 3 fixable with the `--fix` option (8 hidden fixes can be enabled with the `--unsafe-fixes` option).
```
There are 510 errors for PLW2901 and many of those seem like false positives to me so probably at least that rule needs to be disabled. It would be worth investigating the others I think to see if they are generally valid.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -49,6 +49,7 @@ lint.select = [
     "PIE810",
     "PLE",
     "PLR1736",
+    "PLW0602", # Global statement used but no assignment
     "SIM101",
     "SLOT",
     "TRY002",

--- a/sympy/integrals/laplace.py
+++ b/sympy/integrals/laplace.py
@@ -77,7 +77,6 @@ def DEBUG_WRAP(func):
 
 def _debug(text):
     from sympy import SYMPY_DEBUG
-    global _LT_level
 
     if SYMPY_DEBUG:
         print('-LT- %s%s' % ('  '*_LT_level, text), file=sys.stderr)

--- a/sympy/integrals/meijerint.py
+++ b/sympy/integrals/meijerint.py
@@ -568,6 +568,7 @@ def _inflate_fox_h(g, a):
     bs = [(n + 1)/p for n in range(p)]
     return D, meijerg(g.an, g.aother, g.bm, list(g.bother) + bs, z)
 
+
 _dummies: dict[tuple[str, str], Dummy]  = {}
 
 
@@ -588,7 +589,6 @@ def _dummy_(name, token, **kwargs):
     Return a dummy associated to name and token. Same effect as declaring
     it globally.
     """
-    global _dummies
     if not (name, token) in _dummies:
         _dummies[(name, token)] = Dummy(name, **kwargs)
     return _dummies[(name, token)]

--- a/sympy/printing/llvmjitcode.py
+++ b/sympy/printing/llvmjitcode.py
@@ -180,7 +180,7 @@ class LLVMJitCode:
 
     def _create_function_base(self):
         """Create function with name and type signature"""
-        global link_names, current_link_suffix
+        global current_link_suffix
         default_link_name = 'jit_func'
         current_link_suffix += 1
         self.link_name = default_link_name + str(current_link_suffix)
@@ -263,7 +263,6 @@ class LLVMJitCode:
         return vals
 
     def _compile_function(self, strmod):
-        global exe_engines
         llmod = llvm.parse_assembly(strmod)
 
         pmb = llvm.create_pass_manager_builder()

--- a/sympy/simplify/fu.py
+++ b/sympy/simplify/fu.py
@@ -1,6 +1,7 @@
 from collections import defaultdict
 
 from sympy.core.add import Add
+from sympy.core.cache import cacheit
 from sympy.core.expr import Expr
 from sympy.core.exprtools import Factors, gcd_terms, factor_terms
 from sympy.core.function import expand_mul
@@ -653,10 +654,6 @@ def TR10i(rv):
     2*sqrt(2)*x*sin(x + pi/6)
 
     """
-    global _ROOT2, _ROOT3, _invROOT3
-    if _ROOT2 is None:
-        _roots()
-
     def f(rv):
         if not rv.is_Add:
             return rv
@@ -740,7 +737,7 @@ def TR10i(rv):
             # that have the right ratio
             args = []
             for a in byrad:
-                for b in [_ROOT3*a, _invROOT3]:
+                for b in [_ROOT3()*a, _invROOT3()]:
                     if b in byrad:
                         for i in range(len(byrad[a])):
                             if byrad[a][i] is None:
@@ -1718,11 +1715,19 @@ fufuncs = '''
 FU = dict(list(zip(fufuncs, list(map(locals().get, fufuncs)))))
 
 
-def _roots():
-    global _ROOT2, _ROOT3, _invROOT3
-    _ROOT2, _ROOT3 = sqrt(2), sqrt(3)
-    _invROOT3 = 1/_ROOT3
-_ROOT2 = None
+@cacheit
+def _ROOT2():
+    return sqrt(2)
+
+
+@cacheit
+def _ROOT3():
+    return sqrt(3)
+
+
+@cacheit
+def _invROOT3():
+    return 1/sqrt(3)
 
 
 def trig_split(a, b, two=False):
@@ -1775,10 +1780,6 @@ def trig_split(a, b, two=False):
     >>> trig_split(cos(x)*cos(y), sin(x)*sin(y))
     >>> trig_split(-sqrt(6)*cos(x), sqrt(2)*sin(x)*sin(y), two=True)
     """
-    global _ROOT2, _ROOT3, _invROOT3
-    if _ROOT2 is None:
-        _roots()
-
     a, b = [Factors(i) for i in (a, b)]
     ua, ub = a.normal(b)
     gcd = a.gcd(b).as_expr()
@@ -1897,12 +1898,12 @@ def trig_split(a, b, two=False):
         if not cob:
             cob = S.One
         if coa is cob:
-            gcd *= _ROOT2
+            gcd *= _ROOT2()
             return gcd, n1, n2, c.args[0], pi/4, False
-        elif coa/cob == _ROOT3:
+        elif coa/cob == _ROOT3():
             gcd *= 2*cob
             return gcd, n1, n2, c.args[0], pi/3, False
-        elif coa/cob == _invROOT3:
+        elif coa/cob == _invROOT3():
             gcd *= 2*coa
             return gcd, n1, n2, c.args[0], pi/6, False
 

--- a/sympy/testing/tests/diagnose_imports.py
+++ b/sympy/testing/tests/diagnose_imports.py
@@ -123,7 +123,6 @@ if __name__ == "__main__":
     sorted_messages = []
 
     def msg(msg, *args):
-        global options, sorted_messages
         if options.by_process:
             print(msg % args)
         else:
@@ -143,7 +142,6 @@ if __name__ == "__main__":
         question was already imported.
 
         Keeps the semantics of __import__ unchanged."""
-        global options, symbol_definers
         caller_frame = inspect.getframeinfo(sys._getframe(1))
         importer_filename = caller_frame.filename
         importer_module = globals['__name__']

--- a/sympy/utilities/timeutils.py
+++ b/sympy/utilities/timeutils.py
@@ -52,7 +52,6 @@ def _print_timestack(stack, level=1):
 
 def timethis(name):
     def decorator(func):
-        global _do_timings
         if name not in _do_timings:
             return func
 

